src/riscv64/Gresume.c

```diff
@@ -0,0 +1,120 @@
+/* libunwind - a platform-independent unwind library
+   Copyright (C) 2008 CodeSourcery
+
+This file is part of libunwind.
+
+Permission is hereby granted, free of charge, to any person obtaining
+a copy of this software and associated documentation files (the
+"Software"), to deal in the Software without restriction, including
+without limitation the rights to use, copy, modify, merge, publish,
+distribute, sublicense, and/or sell copies of the Software, and to
+permit persons to whom the Software is furnished to do so, subject to
+the following conditions:
+
+The above copyright notice and this permission notice shall be
+included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
+
+
+#include "unwind_i.h"
+#include "offsets.h"
+#include <ucontext.h>
+
+#ifndef UNW_REMOTE_ONLY
+
+HIDDEN inline int
+riscv64_local_resume (unw_addr_space_t as, unw_cursor_t *cursor, void *arg)
+{
+#ifdef __linux__
+  struct cursor *c = (struct cursor *) cursor;
+  ucontext_t *uc = c->uc;
+
+  unw_word_t *mcontext = (unw_word_t*) &uc->uc_mcontext;
+  mcontext[0] = c->dwarf.ip;
+
+  if (c->sigcontext_format == RISCV64_SCF_NONE)
+    {
+      /* Restore PC in RA */
+      mcontext[1] = c->dwarf.ip;
+
+      Debug (8, "resuming at ip=0x%lx via setcontext()\n", c->dwarf.ip);
+
+      setcontext(uc);
+    }
+  else
+    {
+      struct sigcontext *sc = (struct sigcontext *) c->sigcontext_addr;
+      unw_word_t *regs = (unw_word_t*)sc;
+
+      regs[0] = c->dwarf.ip;
+      for (int i = UNW_RISCV64_X1; i <= UNW_RISCV64_F31; ++i) {
+        regs[i] = mcontext[i];
+      }
+
+      Debug (8, "resuming at ip=0x%lx via sigreturn() (trampoline @ 0x%lx, sp @ 0x%lx)\n", c->dwarf.ip, c->sigcontext_pc, c->sigcontext_sp);
+
+      // Jump back to the trampoline
+      __asm__ __volatile__ (
+        "mv x2, %0\n"
```

Comment:
Call it sp not x2, the ABI names are what tools default to and make it more readable